### ISSUE TITLE
Unsync Issue with Translation Mode

### ISSUE THREAD

**Issue by LeonCheung** [SubtitleEdit/subtitleedit#5309]
I believe this is an issue existing for years, though I still tried to figure out how to exactly repeat it, and it occurs in many projects of mine. 

Basically, when you tweaking on either side, the other side should always be synced. However, the problem is that it is not always the case, and it's hard to notice the issue in time until it's a bit too late. :(   (because we only have a single waveform view for either side, not both) Say, even it seems alright in translation, but there may be issues like overlapping or something in original, and it's hard to detect. Even worse, when trying to merge original and translation, it would cause problems by losing something.

Considering that, can we catch the fish here? I mean can we 100% make sure that the adjustment on translation timecodes is also updated to the original? This is very important for bilingual workflow. Thank you!

**Comment by darnn** [comment]
Yeah, I would just like to second this. I'm not sure how you'd go about it, but there should be some way to ensure that the original and the translation always have the same number of subtitles, and that each original subtitle has the same timecodes as the corresponding translation subtitle.
One reproducible way in which this is still broken (at least as of 3.6.1 next beta 331) is this: https://github.com/SubtitleEdit/subtitleedit/issues/5101

**Comment by niksedk** [comment]
I've made an SE test version that will check for sync multiple times each second: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.2/SubtitleEditBetaTranslateSyncCheck.zip

The result is shown in the translate tab: 
![image](https://user-images.githubusercontent.com/1261331/134771664-2e85c6d0-bfd9-43af-8104-67daf015a3b6.png)

I've done a few normal commands, but everything is still in sync...

**Comment by LeonCheung** [comment]
Hmm, I guess maybe it is only likely to occur with certain operations. maybe it's caused by some setting or operations. I'll try this test version in the following days to see if I can finally catch this naughty fish. :)

**Comment by LeonCheung** [comment]
It seems the debug version doesn't work in my case because it always shows OK even though there are indeed something unsync:

![QQ截图20211002185135](https://user-images.githubusercontent.com/494788/135713122-dac15d5b-15b1-40c3-a67e-16796a4a7a50.jpg)

**Comment by niksedk** [comment]
Thx for testing... it was only checking first line... the code in now in latest beta: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.2/SubtitleEditBeta.zip

Set this setting in Settings.xml:    <DebugTranslationSync>True</DebugTranslationSync>

**Comment by darnn** [comment]
Thank you! The method you described to create desync with editing the timing directly using F2 does seem to be fixed here. I'm not quite sure how to test it otherwise since the only ways I knew to reproduce it seem to be fixed, but I'll keep my eye out.  The only thing that still kind of displays this behavior is this:
https://github.com/SubtitleEdit/subtitleedit/issues/5101
But it doesn't seem to be related to desync per se anymore, since it does create the appropriate subtitles with the appropriate timing in both the translation and the original, it's just the splitting itself that doesn't work.

**Comment by LeonCheung** [comment]
Catch it! Finally! It seems only happens when editing the timecodes adjacent to the active line, which is, when editing the end time of the previous line, and/or the start time of the next line. And it seems only happens when the gap between two lines is minimum:

https://user-images.githubusercontent.com/494788/135720874-a908da79-8deb-4361-93c2-8893468e9890.mp4

In the video above, I only used LMB for tweaking and didn't touch any key else.

One interesting thing to note is that when it's out of sync, then I have to hit Ctrl Z several times to go back to "OK" status. During the several Ctrl Zs, you'll notice that during one of the steps, the problematic line seems completely overlapped with the active line.

**Comment by LeonCheung** [comment]
@darnn Would you also help to confirm that? :)

**Comment by darnn** [comment]
I would, but I don't work with waveforms, so I'm not entirely sure what's going on there or how to reproduce it.

**Comment by LeonCheung** [comment]
@darnn, okay, so, you don't work with waveforms, but you still got the out-of-sync issues for unknown reasons right? In that case, there might be more than one naughty fish here... Just try niksedk's debug version and find it out in your own way. :)  Tip: you may want to record the screen while trying to debug it, so that you can play it back to see what's exactly going on when it was out of sync.

**Comment by niksedk** [comment]
Hm, I cannot re-produce the error... any tips?

**Comment by LeonCheung** [comment]
Okay, I just tried again for a while. Forget what I said above. I bet it's related to the setting below:
![image](https://user-images.githubusercontent.com/494788/135726193-d9ac5867-cef4-4e04-b986-8c3649e39ce3.png)

Try testing with a relatively large value.

**Comment by LeonCheung** [comment]
Hmm, so strange that I can still reproduce it even with a value as small as 1,

**Comment by LeonCheung** [comment]
Okay, try this: 
With the factory setting, just try to place the cursor somewhere IN-BETWEEN the gap first, then drag the mouse.

**Comment by niksedk** [comment]
Puh, this is going to be very hard to fix...

**Comment by LeonCheung** [comment]
Hmm, maybe the mechanism should be changed rather than a fix. I mean, when we allow editing translation sub, both translation sub and original sub should share (or be mapped to) the same set of timecodes, rather than trying to sync between the two, what's your idea?

**Comment by niksedk** [comment]
Beta updated: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.6.2/SubtitleEditBeta.zip
How is that?

Yeah, sync'ed original and translation would be easy if they were alike, but sometimes lines are merged or deleted etc. - so this will often not work.

**Comment by LeonCheung** [comment]
Very cool fix, I believe it is well solved! And how about the update issue when working with F2 mode, do you think it needs improving a bit as well?
